core/src/main/java/org/apache/calcite/rel/rules/IntersectToSemiJoinRule.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.rel.rules;
+
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.plan.RelRule;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.core.Intersect;
+import org.apache.calcite.rel.core.JoinRelType;
+import org.apache.calcite.rel.logical.LogicalIntersect;
+import org.apache.calcite.rex.RexBuilder;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.rex.RexUtil;
+import org.apache.calcite.sql.fun.SqlStdOperatorTable;
+import org.apache.calcite.tools.RelBuilder;
+
+import org.immutables.value.Value;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Planner rule that translates a {@link org.apache.calcite.rel.core.Intersect}
+ * to a series of {@link org.apache.calcite.rel.core.Join} that type is
+ * {@link org.apache.calcite.rel.core.JoinRelType#SEMI}.
+ */
+@Value.Enclosing
+public class IntersectToSemiJoinRule
+    extends RelRule<IntersectToSemiJoinRule.Config>
+    implements TransformationRule {
+
+  /** Creates an IntersectToSemiJoinRule. */
+  protected IntersectToSemiJoinRule(Config config) {
+    super(config);
+  }
+
+  //~ Methods ----------------------------------------------------------------
+
+  @Override public void onMatch(RelOptRuleCall call) {
```

Comment:
I am not sure this is correct for multisets.
The result needs to return a row with the MIN of the two counts.

Reply:
this is not about the number of inputs, it's about INTERSECT ALL

Reply:
I don't even know what the right way to implement INTERSECT ALL is. I suspect you will need some aggregates.
But yes, with joins you can only implement INTERSECT.

Reply:
> I don't even know what the right way to implement INTERSECT ALL is. I suspect you will need some aggregates. But yes, with joins you can only implement INTERSECT.

Yes, we cannot know the ndv of the column, which is an inaccurate statistical information. There may be no better way to implement intersect all. I have changed the code to only support intersect.